storage/rocksdb/rdb_io_watchdog.h

```diff
@@ -0,0 +1,113 @@
+/*
+   Copyright (c) 2017, Facebook, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
+
+#pragma once
+
+/* C++ standard header files */
+#include <atomic>
+#include <signal.h>
+#include <stdlib.h>
+#include <string>
+#include <string.h>
+#include <time.h>
+#include <vector>
+
+/* MySQL header files */
+#include "./my_global.h"
+#include "./my_stacktrace.h"
+
+/* MyRocks header files */
+#include "./rdb_utils.h"
+
+namespace myrocks {
+
+class Rdb_io_watchdog {
+  static constexpr int RDB_IO_WRITE_BUFFER_SIZE = 4096;
+  static constexpr const char *const RDB_IO_DUMMY_FILE_NAME = "myrocks_io_watchdog_write_file";
+
+ private:
+  timer_t m_io_check_timer, m_io_check_watchdog_timer;
+  std::atomic<bool> m_io_in_progress;
+  std::vector<std::string> m_dirs_to_check;
+  uint32_t m_write_timeout;
+  mysql_mutex_t m_reset_mutex;
+  char *m_buf;
+
+  int check_write_access(const std::string &dirname) const;
+  void io_check_callback(union sigval timer_data);
+  void expire_io_callback(union sigval timer_data);
+
+  int stop_timers() {
+    int ret = 0;
+
+    if (m_io_check_watchdog_timer) {
+      ret = timer_delete(m_io_check_watchdog_timer);
+
+      if (!ret) {
+        m_io_check_watchdog_timer = nullptr;
+      }
+    }
+
+    if (m_io_check_timer && !ret) {
+      ret = timer_delete(m_io_check_timer);
+
+      if (!ret) {
+        m_io_check_timer = nullptr;
+      }
+    }
+
+    return ret;
+  }
+
+  static void io_check_callback_wrapper(union sigval timer_data) {
+    Rdb_io_watchdog *io_watchdog =
+        static_cast<Rdb_io_watchdog *>(timer_data.sival_ptr);
+    DBUG_ASSERT(io_watchdog != nullptr);
+
+    io_watchdog->io_check_callback(timer_data);
+  }
+
+  static void expire_io_callback_wrapper(union sigval timer_data) {
+    Rdb_io_watchdog *io_watchdog =
+        static_cast<Rdb_io_watchdog *>(timer_data.sival_ptr);
+    DBUG_ASSERT(io_watchdog != nullptr);
+
+    io_watchdog->expire_io_callback(timer_data);
+  }
+
+ public:
+  explicit Rdb_io_watchdog(const std::vector<std::string> &directories)
+      : m_io_check_timer(nullptr), m_io_check_watchdog_timer(nullptr),
+        m_io_in_progress(false), m_dirs_to_check(std::move(directories)),
```

Comment:
Do const ref args and std::move mesh well? The suggested pattern here seems to be pass by value and move std::move one level up, to the constructor caller

Reply:
This is all upstream goodness. I already made several changes to this commit to clean it up and am hesitant to go much further as it will make future auto-merging nearly impossible for the area touched. I _almost_ did not take this commit at all as I saw very little value added for the amount of code it changed and introduced, but, I did not want the same auto-merge issues that would result from the missing commit.

I did not notice the issues with the error handling and that is probably something worth fixing and notifying upstream of the issues or just rolling this commit back out of the merge and taking it in manually in order to correct the issues. I'll have to look closer, review the remaining upstream changes coming and decide whether to keep and fix, just keep as-is, or reject it outright and forward your commentary upstream.

Reply:
All options should result in almost the same cumulative amount of work ultimately

Reply:
Perhaps, but some options allow me to progress through the merging of useful changes and get to fixing real, GA blocking issues while others allow me to punt upstream to fix while accepting some possible short term merge pain.

Reply:
This specific case is interesting, I am surprised that it even compiles as it seems implied that a parameter to std::move must be mutable otherwise no actual move is possible. I suspect there is an implementation detain in std::move that provides a 'const' constructor as well that does no move but a raw copy, which eliminates the entire purpose of even involving std::move here in the first place.

The one place where this is explicitly called, 'directories' is a temporary that is destructed via scope exit right after an instance of Rdb_io_watchdog. It seems to me sending in a list of directories to monitor is construction overkill and the interface could be better defined instead by an 'add_directory' followed by a contract change that reset_timeout must be called after any directory changes have been made.

Reply:
Hmm, that would also produce a poor interface as once the m_directories is 'set' it must be immutable as other state within the implementation depends on the vector remaining unchanged. In fact, due to this, m_directories should be declared as const. Looking at an implementation of std::move, it appears that there are 'const' specifications that allow its use and perform a copy/addref behavior rather than an actual move. So, while its use here is not beneficial, it is also not harmful either.

Reply:
The interface to take directories in the constructor is OK, now the question is how to pass them. 1) The interface could accept only rvalue ref, then std::move in it would be efficient, the only question is again whether that's a good interface. Since it's all very locally used, it could be good enough
2) The interface could not worry about the cost and take the classic const ref, drop the useless std::move then, optimize things later only if the cost is seen to be worth optimizing (probably it won't be). Looks like the current state of the branch.
3) The idomatic way - assuming that std::move on vectors is cheap which AFAIK it is - is to take arg by value and have std::move. When passed an lvalue, the arg is copy constructed, then moved. We have extra move compared to 2), that's OKish (see the "cheap move" requirement above). When passed an rvalue, the arg is move constructed, then moved with std::move. We have an extra move compared to 1), again that's OKish.